tests/store/test_runs_artifact_repo.py

```diff
@@ -0,0 +1,47 @@
+import mock
+import pytest
+
+from mlflow.exceptions import MlflowException
+from mlflow.store.runs_artifact_repo import RunsArtifactRepository
+
+
+@pytest.mark.parametrize("uri, expected_run_id, expected_artifact_path", [
+    ('runs:/1234abcdf1394asdfwer33/path/to/model', '1234abcdf1394asdfwer33', 'path/to/model'),
+    ('runs:/1234abcdf1394asdfwer33/path/to/model/', '1234abcdf1394asdfwer33', 'path/to/model/'),
+    ('runs:/1234abcdf1394asdfwer33', '1234abcdf1394asdfwer33', None),
+    ('runs:/1234abcdf1394asdfwer33/', '1234abcdf1394asdfwer33', None),
+    ('runs:///1234abcdf1394asdfwer33/', '1234abcdf1394asdfwer33', None),
+])
+def test_parse_runs_uri_valid_input(uri, expected_run_id, expected_artifact_path):
+    (run_id, artifact_path) = RunsArtifactRepository.parse_runs_uri(uri)
+    assert run_id == expected_run_id
+    assert artifact_path == expected_artifact_path
+
+
+@pytest.mark.parametrize("uri", [
+    'notruns:/1234abcdf1394asdfwer33/',  # wrong scheme
+    'runs:/',                            # no run id
+    'runs:1234abcdf1394asdfwer33/',      # missing slash
+    'runs://1234abcdf1394asdfwer33/',    # hostnames are not yet supported
+])
+def test_parse_runs_uri_invalid_input(uri):
+    with pytest.raises(MlflowException):
+        RunsArtifactRepository.parse_runs_uri(uri)
+
+
+# We should add an integration test for this code path.
```

Comment:
what would be a good place to add such a test?

Reply:
I think we might be able to write a test here that e.g. creates an experiment & child run, then creates a RunsArtifactRepo against the run (`repo` in the example below) & verifies that the RunsArtifactRepo has the right child ArtifactRepo (`repo.repo.artifact_uri` etc in the example below):

```
>>> mlflow.create_experiment("abc", "s3://abc")
'13'
>>> with mlflow.start_run(experiment_id='13'):
...   active_run_id = mlflow.active_run().info.run_uuid
... 
  
>>> 
>>> active_run_id
'c4ddfe9ae6454f9495dd612da2721653'
>>> from mlflow.store.runs_artifact_repo import RunsArtifactRepository
>>> runs_uri = "runs:/%s/path/to/model" % active_run_id
>>> runs_uri
'runs:/c4ddfe9ae6454f9495dd612da2721653/path/to/model'
>>> repo = RunsArtifactRepository(runs_uri)
>>> repo
<mlflow.store.runs_artifact_repo.RunsArtifactRepository object at 0x115c8bf60>
>>> repo.artifact_uri
'runs:/c4ddfe9ae6454f9495dd612da2721653/path/to/model'
>>> repo.repo
<mlflow.store.s3_artifact_repo.S3ArtifactRepository object at 0x115b5a080>
>>> repo.repo.artifact_uri
's3://abc/c4ddfe9ae6454f9495dd612da2721653/artifacts/path/to/model'
```

Reply:
It happens to work for e.g. S3, but is probably hard to test for artifact repos that do fancy things in their constructors (like the Azure Blob artifact repo)

Reply:
I guess if you really want to integration test that e.g. `RunArtifactRepo.log_artifact` and stuff proxy calls to the run's underlying artifact repo, you could make the experiment/run in the above example use the local FS as its artifact store, rather than S3, and validate that stuff shows up in the local FS after calling `RunArtifactRepo.log_artifact`

Reply:
ok that is much more satisfying. as for integration testing the real methods i'll keep an eye on whether we end up doing it at a higher level anyway - since the risk doesn't seem high here to skip them.